README.md
<div align="center">
  <img src="./img/logo.png">
</div>

[![CircleCI](https://circleci.com/gh/intuit/ReplayZero.svg?style=svg)](https://circleci.com/gh/intuit/ReplayZero)
[![Go version](https://img.shields.io/github/go-mod/go-version/intuit/replayzero)](https://img.shields.io/github/go-mod/go-version/intuit/replayzero)
[![Go Report Card](https://goreportcard.com/badge/github.com/intuit/replayzero)](https://goreportcard.com/report/github.com/intuit/replayzero)
[![License: Apache2.0](https://img.shields.io/badge/License-Apache2.0-blue.svg)](https://opensource.org/licenses/Apache2.0)

A lightweight, localized interface to API recording and test generation. Automatically turn local web traffic from __*your application*__ into functional test scenarios. Replay Zero generates [Karate](https://intuit.github.io/karate/) scenarios by default, but the generation process utilizes a templating engine that allows extension to any HTTP-based test format.

**NOTE:** Replay Zero may be a web proxy, but **DO NOT USE IN PRODUCTION... yet.** This tool was originally made specifically with a developer laptop as the target environment (this can be seen in sample flows below), and should only be viewed as a developer productivity / testing assistant for now.

## Installation

### Homebrew

```sh
brew tap intuit/replay-zero https://github.com/intuit/ReplayZero
brew install replay-zero
```

### Release builds

Binaries are built + attached to releases for Linux, MacOS and Windows.

```sh
# Download a platform specific build
curl -vkL https://github.com/intuit/ReplayZero/releases/download/v0.0.2/replay-zero-osx.zip
# Unpack && rename
unzip replay-zero-osx.zip && mv ./replay-zero-osx ./replay-zero
# Make executable
chmod a+x ./replay-zero
```

## Quickstart

### How does it work

1. HTTP requests are proxied through an embedded web server
1. Request + response data (URI, headers, payloads, response code) are collected and buffered
1. Upon buffer flush, HTTP data is sent through a templating engine and converted into Karate scenarios
1. Test contents are written to files on disk

![Simple recording flow](./img/arch_simple.png)

## Start Recording

1. Start the tool with no arguments, containing the web proxy that listens on `localhost:9000` and forwards (by default) to `localhost:8080`, where your app will run.

```sh
replay-zero
```

1. In another terminal tab, download and start a dummy echo server on port 8080.

```sh
npm install http-echo-server -g
http-echo-server 8080
```

1. Send some requests via curl, specifying Replay Zero as a proxy.

```sh
curl --proxy http://localhost:9000 -H 'my_header: some_value' localhost:8080/sample/api
```

1. Shut down Replay Zero with `Ctrl + C` in the same terminal tab you started it in. Do the same for the echo server if you started that as well.

## Streaming Mode

Instead of saving events locally, Replay Zero can also stream recorded request/response pairs to an [AWS Kinesis data stream](https://aws.amazon.com/kinesis/data-streams/) if you want to set up remote processing/storage.

To enable streaming mode

1. Have valid AWS credentials for your Kinesis stream in either environment variables or a shared credentials file (see FAQ below for more)
1. Pass in values to the following flags
   * `-s` / `--stream-name` - name of Kinesis stream
   * `-r` / `--stream-role-arn` - full IAM role ARN (`arn:aws:iam::<account>:role/...`) for a role that must allow at least the `kinesis:PutRecord` and `kinesis:DescribeStream` actions

### Security

The AWS SDK provides in-transit encryption for API transactions (like the Kinesis `PutRecord` API used to send telemetry). However, Kinesis messages are by default unencrypted at rest while waiting to be consumed from the stream (24 hours by default). Kinesis does offer Server-Side Encryption (SSE) which can be enabled with a default or custom KMS master key.

We recommend enabling Kinesis SSE so that your recorded traffic will not be visible until it is consumed by whatever Kinesis consumer you implement. When provided a valid stream name and IAM role, Replay Zero will call the `kinesis:DescribeStream` API and log a warning if SSE is disabled for that stream (but will still record + send data to the stream).

See the AWS docs on [Kinesis SSE](https://docs.aws.amazon.com/streams/latest/dev/getting-started-with-sse.html) to learn more.

## Additional Features

### Configure Proxy targets

#### Default target port

By default, all events Replay Zero receives will be proxied to `localhost:8080`. The default target port can be changed with the `--target-port` flag

```sh
replay-zero --target-port=8575
```

#### Proxy to multiple targets

By reading the contents of the `Host` header, Replay Zero can rewrite the proxy target on a per-request basis. Certain tools will do this for you (like the above methods for configuring a proxy using cURL or Postman). If you are sending your requests to be proxied some other way, as long as the `Host` header is set with the right localhost target each request will be dynamically rerouted to the right location. Now's a good time to remind that current only localhost / HTTP traffic is supported for proxying.

### Request Batching

Running `replay-zero` with no arguments causes each request/response pair to be written to its own Karate `*.feature` file. But there are several ways to configure consecutive events to be written to the same file.

Things to keep in mind:

* When stopping Replay Zero via `Ctrl+C`, if there are any events left in the buffer they will also be written out so that no data will ever be lost

#### Fixed Batch Size

Pass in an extra `--batch-size <n>` flag that will buffer multiple requests before writing them out. This batch size will be fixed for the duration that Replay Zero is running, and will require a shutdown to change.

#### All-in-one Batch

If you'd like to simply start Replay Zero and have it buffer _everything_ until you shut it down and write all recorded events into 1 big file, pass in any negative batch size:

```sh
replay-zero --batch-size=-1
```

#### Dynamic Batch Size

Say you want to record 2 API flows in your app, one which involves 3 sequential API calls and the other requires 5. Instead of turning on Replay Zero once with `--batch-size=3` and again with `--batch-size=5`, pass an extra header in the *first request* of each flow that looks like `'replay_batch: N'`. Whenever Replay Zero receives a request with this header, it will change its batch size to a valid integer in the header value. After that batch is processed, it will fall back to the default (either the value you passed in `--batch-size` or 1 as a fallback). If a dynamic batch size is received, the event buffer will be flushed to a file so that recorded data will not be lost.

Via cURL, this scenario could look like (with some very complex sequential calls)

```sh
 $ curl -H 'replay_batch: 3' localhost:8080/my/api
hello! here is some data - 123
 $ curl localhost:8080/my/api
hello! here is some data - 456
 $ curl localhost:8080/my/api
hello! here is some data - 789
$ curl -H 'replay_batch: 5' localhost:8080/some/other/api
abc - here is some other data
$ curl localhost:8080/some/other/api
def - here is some other data
$ curl localhost:8080/some/other/api
ghi - here is some other data
$ curl localhost:8080/some/other/api
jkl - here is some other data
$ curl localhost:8080/some/other/api
mno - here is some other data
```

And Replay Zero's output would look something like

```text
Starting nginx + log forwarder
2020/01/16 03:02:33 Running OFFLINE, writing out events to Karate scripts
2020/01/16 03:02:58 Saw event:
a41555fb22543469d126719292390820 GET /my/api
2020/01/16 03:02:58 Collecting dynamic batch size with size 3
2020/01/16 03:03:10 Saw event:
813812b835b4f900f3e6ea7eb7af624a GET /my/api
2020/01/16 03:03:14 Saw event:
5dec9aa40573522fa41966dfbf0e753b GET /my/api
2020/01/16 03:03:14 Wrote 3 scenarios to file replay_scenarios_1.feature
2020/01/16 03:03:35 Saw event:
42c9e477621bc65ac2918ebbfc2f6644 GET /some/other/api
2020/01/16 03:03:35 Collecting dynamic batch size with size 5
2020/01/16 03:03:46 Saw event:
d52734b60a30ba2032edcd4c11e4eb00 GET /some/other/api
2020/01/16 03:03:48 Saw event:
ee87ba117b2ba079dd194b53a8005891 GET /some/other/api
2020/01/16 03:03:49 Saw event:
5330a9b0b1e449e762575bfa97528e4e GET /some/other/api
2020/01/16 03:03:49 Saw event:
e46bfc69b893b659c4db713ba2cc84f1 GET /some/other/api
2020/01/16 03:03:49 Wrote 5 scenarios to file replay_scenarios_2.feature
```

### Different output formats

By default Replay Zero generates Karate `*.feature` files and outputs them to the directory Replay Zero was started in. But the `--template` or `-t` flag allows you to specify the format you'd like your tests to be generated in. The created files will always following the naming format `replay_scenarios_{N}.{extension}`. Out of the box we support below test formats

* Karate  (`*.feature`)
* Gatling (`*.scala`)

Specify the output as a lowercase input to the flag:

```sh
replay-zero --template=gatling
```

#### Custom templates

You can also pass path to your own custom template (in case you dont want to use karate or gatling) to the same paramater and `--extension` or `-e` to pass extentsion of the output.

E.g.

```sh
replay-zero --template=/path/to/your/custom-template --extension=java
```

Templates are following the format from the [text/template](https://golang.org/pkg/text/template/) package. And it also support all template functions provided by [Sprig](http://masterminds.github.io/sprig/). You can even extend your template's functionality with string manipulation, math operators, and more from the Sprig library


## Roadmap

Replay Zero has many plans for improvement which you can find in the Issues tab of this repo. Now that you've read the entire README up until this point (right?) and know of all the features Replay Zero offers, here is a visual recap of both some of the features currently provided (multiple target proxying, output templating) as well as some planned roadmap items (remote proxyingm, custom template sourcing) and how they may fit in alongside existing features.

![Full Feature flow + Roadmap](./img/arch_full.png)

## FAQ

### Q: _My app is running on a port other than 8080._

A: See the "Additional Features" section "Configure Proxy targets" on how to define the target port both globally or per-request.

### Q: _When I run Replay Zero in streaming mode and try to record something, an error about credentials shows up._

(Something like this)

```plain
2019/12/31 00:48:40 NoCredentialProviders: no valid providers in chain
caused by: EnvAccessKeyNotFound: failed to find credentials in the environment.
SharedCredsLoad: failed to load profile, .
EC2RoleRequestError: no EC2 instance role found
caused by: RequestError: send request failed
```

A: Check that either

* you have valid AWS credentials under the right profile in the file `~/.aws/credentials`
* the following environment variables are set in your current shell
  * `AWS_ACCESS_KEY_ID`
  * `AWS_SECRET_ACCESS_KEY`
  * `AWS_SESSION_TOKEN` (optional)

## Telemetry

This tool supports sending usage metrics to an [Amazon Kinesis data stream](https://aws.amazon.com/kinesis/data-streams/), but this is **completely disabled** in the open source repo - there's no URL set to send data to. If you want to use this for internal reporting, set the environment variables

* `REPLAY_ZERO_TELEMETRY_STREAM` (stream name)
* `REPLAY_ZERO_TELEMETRY_ROLE` (IAM role to assume credentials for writing to the stream)

The code for this is in [`telemetry.go`](./telemetry.go) - take a look if you're curious or concerned.

### Security

The AWS SDK provides in-transit encryption for API transactions (like the Kinesis `PutRecord` API used to send telemetry). However, Kinesis messages are by default unencrypted at rest while waiting to be consumed from the stream (24 hours by default). Kinesis does offer Server-Side Encryption (SSE) which can be enabled with a default or custom KMS master key.

We recommend enabling Kinesis SSE so that your telemetry logs will not be visible until it is consumed by whatever Kinesis consumer you implement. When provided a valid stream name and IAM role, Replay Zero will call the `kinesis:DescribeStream` API and log a warning if SSE is disabled for that stream (but will still record + send data to the stream).

See the AWS docs on [Kinesis SSE](https://docs.aws.amazon.com/streams/latest/dev/getting-started-with-sse.html) to learn more.

## Developing

### Requirements

* [Go 1.13.x](https://golang.org/doc/install)
* [golangci-lint](https://github.com/golangci/golangci-lint#install) for linting + static analysis

### Basics

* Running `make` will build the tool for Mac, Linux, and Windows.
* Running `make test` will run tests
* Running `make coverage` will run tests, save the coverage, and then open your browser to the coverage report

### Kinesalite

For both streaming recording and telemetry messages, Replay Zero uses Amazon Kinesis. To test out Kinesis functionality locally, the tool [Kinesalite](https://github.com/mhart/kinesalite) provides a lightweight implementation that works well in a dev environment but can interact with the AWS Kinesis API's.

If you'd like to test your own consumer of either Replay Zero streaming data or telemtry,

1. Follow the Kinesalite docs to install then start it on `https://localhost:4567`
1. When specifying the stream name for Replay Zero set the stream name to `replay-zero-dev` and a special client will be constructed to connect to Kinesalite. Replay Zero will not try to assume an IAM role for credentials with this client.
    * Stream name for HTTP data: `-s` / `--stream-name` flag
    * Stream name for telemetry: `REPLAY_ZERO_TELEMETRY_STREAM` environment variable

## Contribution Guidelines

See the .github/CONTRIBUTING.md file.

## Builds

See the [releases](https://github.com/intuit/replay-zero/releases) page for pre-built binaries.

## License

[Apache 2.0](./LICENSE)


CHANGELOG.md
# v0.1.0 (Mon Oct 26 2020)

#### 🚀 Enhancement

- adding support for custom templates [#22](https://github.com/intuit/ReplayZero/pull/22) ([@shubhamarora](https://github.com/shubhamarora))

#### Authors: 1

- Shubham Arora ([@shubhamarora](https://github.com/shubhamarora))

---

# v0.0.3 (Wed Aug 26 2020)

#### 🐛 Bug Fix

- Use Sprig functions in default templates [#21](https://github.com/intuit/ReplayZero/pull/21) ([@wtait1](https://github.com/wtait1))
- Integrated Sprig Library [#20](https://github.com/intuit/ReplayZero/pull/20) (Siddhi_Mundhra@intuit.com [@siddhi96](https://github.com/siddhi96))

#### 📝 Documentation

- Add Homebrew formula [#19](https://github.com/intuit/ReplayZero/pull/19) ([@wtait1](https://github.com/wtait1))

#### Authors: 3

- [@siddhi96](https://github.com/siddhi96)
- smundhra (Siddhi_Mundhra@intuit.com)
- Will Tait ([@wtait1](https://github.com/wtait1))

---

# v0.0.2 (Mon Mar 30 2020)

#### 🐛 Bug Fix

- Add log levels for Kinesis client [#18](https://github.com/intuit/ReplayZero/pull/18) ([@wtait1](https://github.com/wtait1))
- Remove "go vet" as it is now replaced with golangci-lint [#16](https://github.com/intuit/ReplayZero/pull/16) ([@wtait1](https://github.com/wtait1))

#### 📝 Documentation

- Update contribution development flow [#17](https://github.com/intuit/ReplayZero/pull/17) ([@wtait1](https://github.com/wtait1))

#### Authors: 1

- Will Tait ([@wtait1](https://github.com/wtait1))

---

# v0.0.1 (Tue Mar 24 2020)

#### 🐛 Bug Fix

- Update git info for auto release [#15](https://github.com/intuit/ReplayZero/pull/15) ([@wtait1](https://github.com/wtait1))
- Auto release [#14](https://github.com/intuit/ReplayZero/pull/14) ([@wtait1](https://github.com/wtait1))
- Kinesis telemetry [#13](https://github.com/intuit/ReplayZero/pull/13) ([@wtait1](https://github.com/wtait1))

#### ⚠️  Pushed to `master`

- Update Make target for auto ([@wtait1](https://github.com/wtait1))
- Add "package-lite" Make target that does not require linting ([@wtait1](https://github.com/wtait1))
- Update CircleCI build image ([@wtait1](https://github.com/wtait1))
- Update PULL_REQUEST_TEMPLATE.md ([@Celeo](https://github.com/Celeo))

#### Authors: 2

- Matt Boulanger ([@Celeo](https://github.com/Celeo))
- Will Tait ([@wtait1](https://github.com/wtait1))

main.go
package main

import (
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"strings"
	"syscall"

	"github.com/markbates/pkger"
	flag "github.com/spf13/pflag"

	"github.com/ztrue/shutdown"
)

const (
	verboseCredentialErrors = true
)

var (
	// Version is injected at build time via the '-X' linker flag
	// https://golang.org/cmd/link/#hdr-Command_Line
	Version string

	flags struct {
		version           bool
		listenPort        int
		defaultTargetPort int
		batchSize         int
		template          string
		extension         string
		debug             bool
		streamRoleArn     string
		streamName        string
	}

	client    = &http.Client{}
	telemetry telemetryAgent
)

func check(err error) {
	if err != nil {
		log.Fatal(err)
	}
}

func logErr(err error) {
	if err != nil {
		log.Println(err)
	}
}

func logWarn(msg string, v ...interface{}) {
	if len(v) > 0 {
		println(msg)
		log.Printf("[WARN] "+msg+"\n", v...)
	} else {
		log.Printf("[WARN] " + msg + "\n")
	}
}

func logDebug(msg string, v ...interface{}) {
	if flags.debug {
		if len(v) > 0 {
			log.Printf("[DEBUG] "+msg+"\n", v...)
		} else {
			log.Printf("[DEBUG] " + msg + "\n")
		}
	}
}

func readFlags() {
	// Overrides the default message of "pflag: help requested" in the case of -h / --help
	flag.ErrHelp = errors.New("")
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "Usage of replay-zero:\n")
		flag.PrintDefaults()
	}
	flag.BoolVarP(&flags.version, "version", "V", false, "Print version info and exit")
	flag.IntVarP(&flags.listenPort, "listen-port", "l", 9000, "The port the Replay Zero proxy will listen on")
	flag.IntVarP(&flags.defaultTargetPort, "target-port", "p", 8080, "The port the Replay Zero proxy will forward to on localhost")
	flag.BoolVar(&flags.debug, "debug", false, "Set logging to also print debug messages")
	flag.IntVarP(&flags.batchSize, "batch-size", "b", 1, "Buffer events before writing out to a file")
	flag.StringVarP(&flags.template, "template", "t", "karate", "Either [karate] or [gatling] or [path/to/custom/template]")
	flag.StringVarP(&flags.extension, "extension", "e", "", "For custom output template")
	flag.StringVarP(&flags.streamName, "stream-name", "s", "", "AWS Kinesis Stream name (streaming mode only)")
	flag.StringVarP(&flags.streamRoleArn, "stream-role-arn", "r", "", "AWS Kinesis Stream ARN (streaming mode only)")
	flag.Parse()

	if flags.version {
		fmt.Printf("replay-zero version %s\n", Version)
		os.Exit(0)
	}

	if flags.batchSize == 0 {
		log.Println("Batch size cannot be zero! Using batch=1")
		flags.batchSize = 1
	} else if flags.batchSize > 1 {
		log.Printf("Using fixed batch size of %d\n", flags.batchSize)
	}

	// check if template exist at the provided path
	// TODO: add validation for correctness of template
	if !(flags.template == "karate" || flags.template == "gatling") {
		_, err := ioutil.ReadFile(flags.template)
		if err != nil {
			log.Printf("Failed to load template")
			panic(err)
		}
		if flags.extension == "" {
			log.Fatal("For custom template, output extension is expected to be passed using --extension or -e flag.")
			os.Exit(0)
		}
	}
}

// get embeded template file content
func getPkgTemplate(filePath string) string {
	f, err := pkger.Open(filePath)
	if err != nil {
		panic(err)
	}
	defer f.Close()
	info, err := f.Stat()
	if err != nil {
		panic(err)
	}
	b := make([]byte, info.Size())
	content, err := f.Read(b)
	if err != nil {
		panic(err)
	}
	return string(b[:content])
}

func getFormat(template string, extension string) outputFormat {
	switch template {
	case "karate":
		return outputFormat{
			template:  getPkgTemplate("/templates/karate_default.template"),
			extension: "feature",
		}
	case "gatling":
		return outputFormat{
			template:  getPkgTemplate("/templates/gatling_default.template"),
			extension: "scala",
		}
	default:
		dat, err := ioutil.ReadFile(template)
		if err != nil {
			panic(err)
		}
		return outputFormat{
			template:  string(dat),
			extension: extension,
		}
	}
}

func buildNewTargetURL(req *http.Request) string {
	scheme := req.URL.Scheme
	if scheme == "" {
		scheme = "http"
	}

	host := req.URL.Host
	if host == "" {
		host = fmt.Sprintf("localhost:%d", flags.defaultTargetPort)
	}
	return fmt.Sprintf("%s://%s%s", scheme, host, req.URL.Path)
}

// A higher-order function that accepts an HTTPEvent handler and
// returns a network interceptor that passes parses + builds an HTTPEvent
// and passes that on to the parametrized handler for further processing.
func createServerHandler(h eventHandler) func(http.ResponseWriter, *http.Request) {
	return func(wr http.ResponseWriter, originalRequest *http.Request) {
		// 1. Construct proxy request
		newURL := buildNewTargetURL(originalRequest)
		defer originalRequest.Body.Close()
		originalBody, err := ioutil.ReadAll(originalRequest.Body)
		if err != nil {
			log.Printf("[ERROR] Could not read original request body: %v\n", err)
			return
		}
		originalBodyString := string(originalBody)
		request, err := http.NewRequest(originalRequest.Method, newURL, strings.NewReader(originalBodyString))
		if err != nil {
			log.Printf("[ERROR] Could not build the outgoing request: %v\n", err)
			return
		}
		request.Header = originalRequest.Header

		// 2. Execute proxy request
		response, err := client.Do(request)
		if err != nil {
			log.Printf("[ERROR] Could not process HTTP request to target: %v\n", err)
			return
		}

		// 3. Copy data for proxy response
		for k, v := range response.Header {
			wr.Header().Set(k, strings.Join(v, ","))
		}
		wr.WriteHeader(response.StatusCode)
		defer response.Body.Close()
		originalRespBody, err := ioutil.ReadAll(response.Body)
		if err != nil {
			log.Printf("[ERROR] Could not read response body: %v\n", err)
			return
		}
		_, err = io.Copy(wr, strings.NewReader(string(originalRespBody)))
		if err != nil {
			log.Printf("[ERROR] Could not create copy of response body: %v\n", err)
			return
		}
		originalRespBodyString := string(originalRespBody)
		// 4. Parse request + response data and pass on to event handler
		event, err := convertRequestResponse(request, response, originalBodyString, originalRespBodyString)
		if err != nil {
			log.Printf("[ERROR] Could not convert request and response: %v\n", err)
			return
		}

		log.Printf("Saw event:\n%s %s %s", event.PairID, event.HTTPMethod, event.Endpoint)
		h.handleEvent(event)
	}
}

func main() {
	readFlags()
	telemetry = getTelemetryAgent()
	go telemetry.logUsage(telemetryUsageOpen)

	var h eventHandler
	if len(flags.streamName) > 0 {
		if len(flags.streamRoleArn) == 0 {
			log.Println("AWS Kinesis Stream ARN and name required for streaming mode")
			os.Exit(1)
		}
		log.Println("Running ONLINE, sending recorded events to Kinesis")
		h = getOnlineHandler(flags.streamName, flags.streamRoleArn)
	} else {
		log.Printf("Running OFFLINE, writing out events to %s files\n", flags.template)
		h = getOfflineHandler(flags.template, flags.extension)
	}

	shutdown.Add(func() {
		log.Println("Cleaning up...")
		h.flushBuffer()
	})

	http.HandleFunc("/", createServerHandler(h))
	listenAddr := fmt.Sprintf("localhost:%d", flags.listenPort)
	log.Println("Proxy listening on " + listenAddr)
	go func() {
		log.Fatal(http.ListenAndServe(listenAddr, nil))
	}()

	shutdown.Listen(syscall.SIGINT)
}
